### ISSUE TITLE
Memory leak?  

### ISSUE THREAD

**Issue by tobidelbruck** [swri-robotics/bag-reader-java#32]
When reading in BagFile I got this error, not sure if it should occur or not. The bag file is about 2GB in size, but only one message is read and stored at any one moment from the output of getMessageOnTopicAtIndex; i.e. jaer resuses the objects it stores the messages to. The exception is thrown at line 307 of https://github.com/swri-robotics/bag-reader-java/blob/master/src/main/java/com/github/swrirobotics/bags/reader/messages/serialization/MessageType.java  when the timeStats stores the new deltaTime

       timeStats.add(finishTimeNs - startTimeNs);
 
````
WARNING: java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3181)
	at java.util.ArrayList.grow(ArrayList.java:261)
	at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:235)
	at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:227)
	at java.util.ArrayList.add(ArrayList.java:458)
	at com.github.swrirobotics.bags.reader.messages.serialization.MessageType.readMessage(MessageType.java:307)
	at com.github.swrirobotics.bags.reader.messages.serialization.ArrayType.readMessage(ArrayType.java:284)
	at com.github.swrirobotics.bags.reader.messages.serialization.MessageType.readMessage(MessageType.java:294)
	at com.github.swrirobotics.bags.reader.BagFile.getMessageOnTopicAtIndex(BagFile.java:599)
	at net.sf.jaer.eventio.ros.RosbagFileInputStream.getNextMsg(RosbagFileInputStream.java:147)
	at net.sf.jaer.eventio.ros.RosbagFileInputStream.getNextRawPacket(RosbagFileInputStream.java:178)
	at net.sf.jaer.eventio.ros.RosbagFileInputStream.readPacketByTime(RosbagFileInputStream.java:235)
	at net.sf.jaer.graphics.AEPlayer.getNextPacket(AEPlayer.java:472)
	at net.sf.jaer.graphics.AEViewer$ViewLoop.grabInput(AEViewer.java:2092)
	at net.sf.jaer.graphics.AEViewer$ViewLoop.run(AEViewer.java:1816)
````

**Comment by pjreed** [comment]
Hmm, I didn't think that gathering those stats would use so much memory, but maybe that is the case.  If you try commenting out that block (basically from line 296 through 307), do you still have the issue?

**Comment by cheerss** [comment]
1. The bag file is read by using java.nio.channels.FileChannel, but the file is never closed.
2. Every time we execute BagReader.readfile(String filename), all data in the bag file would be read into the memory.
So do these lead to memory leak?

**Comment by pjreed** [comment]
I don't think that's the case; everywhere a FileChannel is opened, it's done in a try-with-resources block, which is guaranteed to close the FileChannel when the block exits.

I think the "leak" was actually due to a map that was being used to collect stats on message parsing but was never being cleared;  it should be fixed in in PR #33.

I intend to merge that in soon, I just want to make a few coding style tweaks and test it myself, and I haven't had time to do so yet.

**Comment by cheerss** [comment]
Thank you very much for answering my question. I found that my program crushes because of the same reason as "Pull Request #33", I could only download the code and repair it as tobidelbruck says and using my own .jar. I would appreciate it if you merging the request.

**Comment by pjreed** [comment]
Fixed in #33, so closing this.
